### PR TITLE
[FLINK-32468][rpc] Switch from Akka to Pekko

### PR DESCRIPTION
**Pull request by zentol**
Switch to the Apache fork of Akka. Akka support ends in September, so 1.18.x would eventually be affected.

I decided not to rename classes/modules because it feels like a lot of noise in the git history for virtually no gain.
May also make it easier to revert things if we do run into a problem during release testing.

### CODE REVIEW

File: flink-rpc/flink-rpc-akka/src/main/resources/META-INF/NOTICE
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/resources/META-INF/NOTICE]: ~RC2 needs to be removed~

File: flink-rpc/flink-rpc-akka/src/main/resources/META-INF/NOTICE
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/resources/META-INF/NOTICE:11]: Were these other dependencies updated as transitive dependencies of Pekko? :thinking:

**Review by XComp — CHANGES_REQUESTED** [review]
`flink-rpc/flink-rpc-akka-loader`:pom.xml -> "Prevent akka and scala from being visible ..."
`flink-rpc/flink-rpc-akka`:pom.xml -> "For dependency convergence in Akka 2.6.20"
`AkkaBasedEndpoint` JavaDoc: "Interface for Akka based rpc gateways."
`AkkaInvocationhandler` JavaDoc: "The Akka (RPC) address of {@link #rpcEndpoint} ..."
`AkkaInvocationHandler` log: "... This is usually caused by: 1) Akka failed sending ..."
`PriorityThreadsDispatcher` JavaDoc: "Akka dispatcher threads creates..."
`RemoteAddressExtension` JavaDoc: "{@link akka.actor.ActorSystem} {@link Extension} used ..."
`RobustActorSystem` comment: "some parts of the akka shutdown procedure ..."
`flink-table/flink-table-planner-loader/pom.xml`: "Prevent akka and scala from being visible to ..."

I guess, the comments should be properly converted, though. Don't you think?

File: flink-rpc/flink-rpc-akka/src/main/resources/META-INF/NOTICE
Comment (zentol) [reply flink-rpc/flink-rpc-akka/src/main/resources/META-INF/NOTICE:11]: yes. Pekko 1.0.0 is not a pure fork of Akka 2.6.20 unfortunately.

**Comment by XComp** [comment]
>I decided not to rename classes/modules because it feels like a lot of noise in the git history for virtually no gain.
May also make it easier to revert things if we do run into a problem during release testing.

I see your point. But in my opinion, there are even more examples (besides the ones I mentioned in my previous comment) where we should switch to pekko, e.g. see [AkkaOptions:361](https://github.com/apache/flink/blob/e3609257687ddb726a1bbca4c5954b0714d2986a/flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java#L361]) which should actually point to https://pekko.apache.org/docs/pekko/current/remoting-artery.html#failure-detector)

**Comment by zentol** [comment]
I did another pass over all references containing akka that aren't modules/classes/variables; this was mostly cosmetic but there were also genuine issues.
I also renamed all akka related config keys (with deprecated akka keys of course)

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java:466]:
```suggestion
     * Creates a pekko config with the provided configuration values. If the listening address is
```
nit

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java:471]:
```suggestion
     *     null is given, then a Pekko config for local actor system will be returned
```
nit

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java:452]:
```suggestion
     *     given, then a Pekko config for local actor system will be returned
```
nit: since you fixed it properly in the JavaDoc above :man_shrugging:

File: flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniClusterConfiguration.java
Comment (XComp) [inline flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniClusterConfiguration.java:92]: I guess it makes sense to remove the prefix :+1:

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java
Comment (XComp) [inline flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java:44]:
```suggestion
     * <b>HACK:</b> Set to 21474835 seconds, Pekko's maximum delay (Pekko 1.0.0). The value cannot
```

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/DeadLettersActor.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/DeadLettersActor.java:1]: The JavaDoc refers to the Akka class `DeadLetter` but should rather point to the corresponding Pekko class.

File: flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:336]:
```suggestion
                                                    + " Pekko’s DeathWatch can be found %s",
```

File: flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:358]:
```suggestion
                                                    + " Higher value increases the time to detect a dead TaskManager. A thorough description of Pekko's"
```

File: flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:379]:
```suggestion
                                                    + " a high value increases the time to detect a dead TaskManager. A thorough description of Pekko's"
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/net/ConnectionUtils.java
Comment (XComp) [inline flink-runtime/src/main/java/org/apache/flink/runtime/net/ConnectionUtils.java:422]:
```suggestion
                                    "Retrieved new target address {} for RPC URL {}.",
```

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/PriorityThreadsDispatcher.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/PriorityThreadsDispatcher.java:1]: The constructors JavaDoc has two occurrences of the `Akka` term.

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java:143]: That's a bit off-topic (hotfix) but could we add a JavaDoc to this method that mentions `PriorityThreadsDispatcher`? Initially, I got confused by Intellij telling me that `PriorityThreadsDispatcher` is an "unused" class. Adding a descriptive JavaDoc here would make this issue go away.

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcService.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcService.java:83]:
```suggestion
 * Pekko based {@link RpcService} implementation. The RPC service starts an actor to receive
```

File: flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:72]:
```suggestion
    /** Timeout for pekko ask calls. */
```
It might be reasonable to modify the JavaDoc here (and the other AkkaOptions entries) as well. ...since we're touching JavaDoc in other cases as well.

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java
Comment (XComp) [inline flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java:47]:
```suggestion
     * org.apache.pekko.actor.LightArrayRevolverScheduler}.
```
That's odd: My Intellij isn't able to resolve the reference even though it's the correct location. :thinking: It even suggests this class name when trying to fix the error

File: flink-rpc/flink-rpc-akka/src/test/java/org/apache/flink/runtime/rpc/akka/ContextClassLoadingSettingTest.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/test/java/org/apache/flink/runtime/rpc/akka/ContextClassLoadingSettingTest.java:1]: nit: JavaDoc contains "akka" twice

**Review by XComp — CHANGES_REQUESTED** [review]
Renaming everything would make the review easier. :innocent: 

Anyway, I'm wondering whether we could add a README to the `flink-rpc-akka` module where we could explain the history behind the Akka classes and Pekko documentation. But I understand that this could be also covered by the git history. I just think that having akka and pekko in the codebase side by side might be confusing to readers in the future.

Other code location I found:
* There is a config parameter `TaskManagerOptions.EXIT_ON_FATAL_AKKA_ERROR` that contains `akka`
* `ActorSystemExtension` and `AkkaActorSystemTest` have invalid references in its JavaDoc
* `RpcEndpoint`'s JavaDoc mentions Akka
* nit: `MessageSerializationTest` mentions akka

**Comment by zentol** [comment]
> There is a config parameter TaskManagerOptions.EXIT_ON_FATAL_AKKA_ERROR that contains akka

Deprecated.

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java
Comment (zentol) [reply flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java:44]: I dont know if that is the maximum delay. I'd keep it as the source of the original value.

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/PriorityThreadsDispatcher.java
Comment (zentol) [reply flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/PriorityThreadsDispatcher.java:1]: due to packages

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java
Comment (zentol) [reply flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaUtils.java:143]: fixed in another way

File: flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java
Comment (zentol) [reply flink-rpc/flink-rpc-core/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java:47]: Expected because it's not a public class.

File: flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java
Comment (XComp) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:80]: Why did that change?

**Review by XComp — CHANGES_REQUESTED** [review]
I couldn't find any other occurrences except for the `AkkaOptions` which is `@PublicEvolving`. Shall we add a deprecation flag here and plan to rename it with 2.0? A spotless:apply run is missing. But it looks good otherwise.

On the other note: Have we thought of providing a separate module `flink-rpc-pekko` and providing that one along `flink-rpc-akka` in 1.18 with pekko being the default one? Or is this too much of an effort and we're comfortable enough that Pekko works as is? 

And we might want to brief the 1.18 release managers on that PR considering that we missed the feature freeze for 1.18 last weekend.

**Comment by zentol** [comment]
> Have we thought of providing a separate module flink-rpc-pekko and providing that one along flink-rpc-akka in 1.18 with pekko being the default one? 

Yes, but that implies duplicating all classes and increasing the dist size by ~30mb.

> Or is this too much of an effort and we're comfortable enough that Pekko works as is?

I wouldn't say I'm comfortable, but I think it's better to potentially break things now than doing this in a bugfix release or running with an unsupported Akka version,

> Shall we add a deprecation flag here and plan to rename it with 2.0?

That's a separate discussion I believe because it's an API breaking change.

**Comment by XComp** [comment]
The changes look good now. I will bring this PR up in tomorrows 1.18 release sync. Would we break anything if we also rename pom modules? ...now that we've renamed the packages and classes as well.

**Comment by zentol** [comment]
> Would we break anything if we also rename pom modules?

I think it'd be annoying for devs, because that implies having to rebuild the rpc system whenever you switch between the pre and post pekko state of the project.

**Review by XComp — APPROVED** [review]
That's a good point. Nothing to add from my side. :+1: Let's wait for CI to become green and the release sync tomorrow.

**Comment by XComp** [comment]
fyi: the e2e1 test failure is fixed with FLINK-32632

File: flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java
Comment (mdedetrich) [inline flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:1]: This file is still called `AkkaOptions.java`, shouldn't it be renamed to `PekkoOptions.scala`?

Comment (mdedetrich) [reply flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:1]: Disregard comment, didn't realize its public API.

Comment (XComp) [reply flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java:1]: AkkaOptions is left like that because it's labeled as `@PublicEvolving`. Deprecating this one requires a discussion which we would do in a separate step. But thanks for looking over it. :+1:

**Comment by zentol** [comment]
We can still bump it in the following days if needed.

**Comment by pjfanning** [comment]
I will release Pekko 1.0.1 jars early tomorrow (UTC). They are up for 24 hours for review. The pekko-remote issue is in an internal class but it is better to use thew release.

**Comment by pjfanning** [comment]
@zentol Pekko 1.0.1 is available now

**Comment by mdedetrich** [comment]
PR that bumps Pekko from 1.0.0 to 1.0.1 has been created at https://github.com/apache/flink/pull/23080

**Comment by He-Pin** [comment]
Hi team, I want to upgrade the current classic transport from netty 3 to netty 4 ,do you think that's ok?
I just asked on maillist https://lists.apache.org/thread/grzp3jw398rtpc6oqfybzxnry2mtyh64

**Comment by pjfanning** [comment]
Pekko 1.0.x will be maintained for a while. It is only Pekko 1.1.x where we are considering changing to netty 4 (or even dropping netty and just keeping Artery remoting).

**Pull request merged**
